linux-jovian+gamescope: Update to 20230516.1000 feature level

samueldr opened the pull request:
This updates the kernel and gamescope to the version used in ~~20230426.1000~~ 20230516.1000.

The update needs to happen together, or things don't work exactly well. It also requires adding the new color feature for complete support.

Note that I personally am using it paired with #78.

* * *

#### What was done

 - Launched steam through a systemd unit, verified that saturation knob works.
 - Verified that logs (journalctl) aren't spewing loads of crap
 - :new: Verified colours are fine (not weirdly dark) in desktop mode
 - :new: Verified "night mode" (steam's redshift) now works fine again
 - :new: Verified gamescope+steam works with these changes on mainline (6.3)
 - :new: Verified state of the "current" (previous) vendor kernel version...

#### Known problems (unsolvable)

~~Users updating to this state, but who will not reboot, will be stranded without the gamescope interface when/if it is restarted.~~

~~We *could* attempt to sniff at the kernel, and when it is lower than `6.1.21-valve1` *do something* at runtime, but what exactly is hard to decide. We can't rely on gamescope to run something to notify the user!~~

I forgot about `--disable-color-management`. With `--disable-color-management` things are now correct.

We can check for `6.1.*-valve*` and tack on `--disable-color-management` for now. We don't have to do anything fancy for mainline 6.3.

NOTE: I tested this by making this local change with `git checkout jovian/development -- ./pkgs/linux-jovian/`.

Verifying this applies correctly can be done with `pgrep -fla /run/wrappers/bin/gamescope`.

samueldr committed:
linux-jovian: Add workaround for vendor oopsies

samueldr committed:
linux-jovian: 6.1.21-valve1 -> 6.1.21-joshcolor1

This was shipped in 20230426.1000

samueldr committed:
modules/steam: Add new misc. environment variables

samueldr committed:
modules/steam: Add new gamescope environment variables

samueldr committed:
gamescope: Init at 3.12.0-beta2

Co-authored-by: Zhaofeng Li <hello@zhaofeng.li>

samueldr commented on pkgs/linux-jovian/default.nix:
```diff
@@ -109,6 +109,19 @@ buildLinux (args // rec {
-    hash = "sha256-ypYhz1kD+enIl31yhjlzqDnd3RqQcyc+7Udlw1Y5iSM=";
+    hash = "sha256-nwog+yXv9v1vENLELwJrtizRv2w7PVWivc+NvAL+Gn0=";
+
+    # Sometimes the vendor doesn't update the EXTRAVERSION tag.
+    # Let's fix it up in post.
+    # ¯\_(ツ)_/¯
+    # Also, `postPatch` on the kernel doesn't compose in `buildLinux`.
+    # ¯\_(ツ)_/¯
+    postFetch = ''
+      (
+      echo ":: Fixing-up EXTRAVERSION with actual tag"
+      cd $out
+      sed -i -e 's/^EXTRAVERSION =.*/EXTRAVERSION = -${vendorVersion}/g' Makefile
+      )
+    '';
```
This mess is thanks to the vendor not updating `EXTRAVERSION` with this particular tag update.

I preferred making our version match the actual tag rather than match the vendor source.

This can be kept going forward as either it'll be a no-op (whitespace notwithstanding) or will fix issues that *anyway* our build tooling will be angry at.

Namely, without doing *some other acrobatics*, we need to have the package version match the "mod dir", which in turn defaults to `${version}`, which includes `vendorVersion`.

samueldr commented on modules/steam.nix:
```diff
@@ -340,6 +340,9 @@ in
         # Disable automatic audio device switching in steam, now handled by wireplumber
         STEAM_DISABLE_AUDIO_DEVICE_SWITCHING = "1";
 
+        # Let steam know it can unmount drives without superuser privileges
+        STEAM_ALLOW_DRIVE_UNMOUNT = "1";
+
```
I've not seen any different behaviour from Steam with this on. Vendor added this previously.

I suppose Steam has to handle the mounting to get this going? Or maybe it needs to be mounted using udisks? My SD card is auto-mounted by systemd on access [assuming the right one is inserted].

samueldr commented on modules/steam.nix:
```diff
@@ -403,6 +415,9 @@ in
         # Enable dynamic backlight, we have the kernel patch to disable events
         STEAM_ENABLE_DYNAMIC_BACKLIGHT = "1";
 
+        # Enabled fan control toggle in steam
+        STEAM_ENABLE_FAN_CONTROL = "1";
+
```
I don't know what this adds. Vendor added this since I last looked, no obvious toggles in Steam. Might be dead code.

samueldr commented on modules/steam.nix:
```diff
@@ -364,6 +367,15 @@ in
         # Enable VRR controls in steam
         STEAM_GAMESCOPE_VRR_SUPPORTED = "1";
 
+        # Scaling support
+        STEAM_GAMESCOPE_FANCY_SCALING_SUPPORT = "1";
+
```
Unsure what this does... It might be adding a slider in the quick access menu to scale the interface when connected to an external display.

samueldr replied:
- https://www.steamdeckgaming.net/post/steam-os-3-5-upcoming-changes

Under *New Scaling Options*, NIS scaling and more interestingly, *Scaling Mode* where the image can be "fit", "stretch" or "fill". ~~The latter I assume is like `background-size: cover;` in CSS.~~ Confirmed, it ensures there's no black borders, but will crop the image.

You can test by forcing a game to e.g. `640x480` resolution, and then it starts all making sense.

If you didn't know, in a games' properties you can set the resolution it should be ran as, *Native* is useful when playing things that the Deck can run natively on an external display, like an emulator. Make sure to (re)start the game with the external display on. The default is AFAIUI to run at the deck's internal resolution.

zhaofengli commented:
With the new kernel, Plasma on X11 now looks very ~~unsaturated~~ dark on the internal display. Not sure if this is also a problem on vanilla SteamOS.

samueldr commented:
I was asked to test on non-joshcolor:

```
Apr 29 15:47:02 dashdingo games-session-start[1886]: xwm: We failed our modeset and have no mode to fall back to! (Initial modeset failed?): Invalid argument                                  
Apr 29 15:47:02 dashdingo systemd[1]: Started Process Core Dump (PID 2137/UID 0).                                                                                                              Apr 29 15:47:03 dashdingo systemd-coredump[2138]: [🡕] Process 1886 (gamescope-wl) of user 1000 dumped core.                                                                                                                                                                                                                                                                                                                                     Module libpipewire-module-session-manager.so without build-id.                                                                               
```

Which smells like

 - https://github.com/ValveSoftware/gamescope/commit/5af321724c8b8a29cef5ae9e31293fd5d560c4ec

(Updating to beta3)

EDIT: No it's not that.

samueldr commented:
Preliminary look at the situation: using gamescope with the previous "actual" `6.1.21-valve1` fails, even with the color stuff disabled as a flag.

samueldr commented:
```
faee3be9def04a0493a4cb6359ce5b6b9202e1b1 is the first bad commit
commit faee3be9def04a0493a4cb6359ce5b6b9202e1b1
Author: Joshua Ashton <joshua@froggi.es>
Date:   Fri Mar 31 11:32:15 2023 +0100

    drm: Move to per-plane 3D LUT + Shaper
    
    Need to do some cleanup here and version checks before merging.

 src/drm.cpp          | 131 ++++++++++++++-------------------------------------
 src/drm.hpp          |   3 +-
 src/steamcompmgr.cpp |   2 +-
 3 files changed, 37 insertions(+), 99 deletions(-)
```

 - https://github.com/ValveSoftware/gamescope/commit/faee3be9def04a0493a4cb6359ce5b6b9202e1b1

(Which has a known workaround of using `--disable-color-management`)

samueldr committed:
modules/steam: Add gamescope EDID patch paths

Using the same path as the vendor, such that if it's handled by
gamescope itself down the line, things will continue working.

samueldr commented:
So, using `--disable-color-management` fixes the crash on non-joshcolor kernels, but the colours are all wrong in the running Steam.

For now let's pause until vendor does *something*.

Using the whole PR in a coherent update + reboot is safe on Steam Deck. Usage on other hardware is untested.

For reference, colours being all wrong is not about the hue. It's about the levels. Grays are crushed to absolute black.

samueldr commented:
In the Matrix room:

```
K900
I think I got the weird colors thing too
K900
On SteamOS
samueldr
Ooh, thanks, that helps 
```

Assuming this is talking about desktop mode colours being as screwey as they are for us.

samueldr committed:
linux-jovian: 6.1.21-joshcolor1 -> 6.1.21-joshcolor2

samueldr committed:
gamescope: 3.12.0-beta2 -> 3.12.0-beta5

samueldr committed:
modules/steam: Add workaround for color management and vendor kernels

samueldr commented:
This is now ready to review. Testing with previous kernel and mainline is welcome, but not strictly necessary.

samueldr commented:
Making a note here, the last kernel update is uh, not needed to be really checked deeply

  - https://github.com/Jovian-Experiments/linux/compare/6.1.21-joshcolor2...6.1.21-saml

We can either ignore it altogether, or bundle the change.

zhaofengli committed:
gamescope: 3.12.0-beta5 -> 3.12.0-beta8

zhaofengli committed:
linux-jovian: 6.1.21-joshcolor2 -> 6.1.21-saml

zhaofengli commented:
> We can either ignore it altogether, or bundle the change.

Let's update while we are at it. I've also bumped gamescope which seems to work fine.

samueldr commented on pkgs/gamescope/default.nix:
```diff
@@ -0,0 +1,37 @@
+, gbenchmark
+}:
+
+# NOTE: vendoring gamescope for the time being since we want to match the
+#       version shipped by the vendor, which does not work without the kernel
+#       version coherent with the version as shipped by the vendor.
+
```
This comment is now partially untrue... We may want to rephrase it so it's only about keeping the version coherent with the shipped vendor version. The issues with it working only with the vendor kernel has been resolved since.

(I'll handle it later today.)

samueldr committed:
gamescope: Fix comment about matching vendor version

samueldr commented:
With the workaround for the issue launching steam in use, I was able to verify steam does launch as it used to.

 - https://github.com/Jovian-Experiments/Jovian-NixOS/issues/94

(Cross-referencing in case anyone else tests, thus updates their steam, and thus breaks their steam.)
